dvc/stage.py

```diff
@@ -565,7 +569,12 @@ def load(repo, fname):
         Stage._check_dvc_filename(fname)
         Stage._check_isfile(repo, fname)
 
-        d = load_stage_file_fobj(repo.tree.open(fname), fname)
+        # NOTE: closing() is only needed for Python 2.7
```

Comment:
Is leaking fds not considered bad in Python 3? :slightly_smiling_face:

Reply:
In Python 3 `open()` is a context manager itself.

Reply:
When dropping Python 2 we will just remove `closing()` leaving `with` block there. Comments like this will help easily find code to change.

Reply:
@Suor Maybe we could return `closing()` in the `tree` itself then and leave a note about it there, instead of copying it everywhere?

Reply:
Using closing in a tree will prevent normal use of `.open()`. We may provide our subclasses of `StringIO` and `BytesIO` for Python 2 to make this transparently work everywhere, but is not worth the effort.

Reply:
@Suor Should be pretty simple to wrap them in `compat.py`, or am I missing something? At least we won't have to go looking for "is only needed for Python 2.7" comments and will have everything contained within `compat.py`.

Reply:
The value of this change is around zero, we just hide a bit of complexity in other file, we grow overall complexity so I say it has negative value. It might had payed off if we used `.tree.open()` frequently, but most probably we will have those two uses next 8 months only.

So the best strategy is to wait 8 months and then remove this code with Python 2.7 support.

Reply:
@Suor Again, that other file is a special file where we hide all py2/3 compatibility wrappers. Removing it is much easier than going around searching for your duplicated comments around the code, that don't even mention anything about StringIO. How do you imagine anyone would read your ` # NOTE: closing() is only needed for Python 2.7` comment in a year and understand that it was about StringIO not having __exit__ in python2? I'm not sure what we are fighting for, is it so hard to add
```
class StringIO(StringIO.StringIO):
    def __enter__(self):
        return self

    def __exit__(self, *args):
        self.close()
```
to `compat.py` and be done with it?

Reply:
This all might be right in the long run, but there won't be. Right now this is useless or negative value change, which should not be even discussed.

As for comment, anyone can search through project for "Python 2" and "Python 3" and remove `closing()`, he won't need to understand anything about StringIO. And  whoever will make Python 2 drop will need to run that search anyway and that will take less time than doing senseless changes now.

Reply:
@Suor 1 year is a long enough run to not create problems for ourselves by polluting code with nondescriptive comments. This is why when some piece of code is used more than once, in all other places we use `compat.py` and not copypasting `NOTE: only needed for Python2` all over, which we would have to go looking for later. We do need to understand that `closing` is only needed for `StringIO`, because from your current comment it is not clear and people might get an impression that there are other things that need that. By adding a StringIO wrapper we mitigate the problem once and for all, and if anyone uses `tree.open` later he won't have to be aware of having to use `closing()` with it, because our wrapped StringIO would already be compatible with `with`.